Content.Shared/Ninja/Components/NinjaGlovesComponent.cs

```diff
@@ -0,0 +1,151 @@
+using Content.Shared.Actions;
+using Content.Shared.Actions.ActionTypes;
+using Content.Shared.DoAfter;
+using Content.Shared.Ninja.Systems;
+using Content.Shared.Tag;
+using Content.Shared.Toggleable;
+using Robust.Shared.Audio;
+using Robust.Shared.GameStates;
+using Robust.Shared.Serialization;
+using Robust.Shared.Serialization.TypeSerializers.Implementations.Custom.Prototype;
+using Robust.Shared.Utility;
+using System.Threading;
+
+namespace Content.Shared.Ninja.Components;
+
+/// <summary>
+/// Component for toggling glove powers.
+/// Powers being enabled is controlled by User not being null.
+/// </summary>
+[RegisterComponent, NetworkedComponent, AutoGenerateComponentState]
+[Access(typeof(SharedNinjaGlovesSystem))]
+public sealed partial class NinjaGlovesComponent : Component
+{
+    /// <summary>
+    /// Entity of the ninja using these gloves, usually means enabled
+    /// </summary>
+    [ViewVariables, AutoNetworkedField]
+    public EntityUid? User;
+
+    /// <summary>
+    /// The action for toggling ninja gloves abilities
+    /// </summary>
+    [DataField("toggleAction")]
+    public InstantAction ToggleAction = new()
+    {
+        DisplayName = "action-name-toggle-ninja-gloves",
+        Description = "action-desc-toggle-ninja-gloves",
+        Priority = -13,
+        Event = new ToggleActionEvent()
+    };
+}
+
+/// <summary>
+/// Component for emagging doors on click, when gloves are enabled.
+/// Only works on entities with DoorComponent.
+/// </summary>
+[RegisterComponent]
+public sealed class NinjaDoorjackComponent : Component
+{
+    /// <summary>
+    /// The tag that marks an entity as immune to doorjacking
+    /// </summary>
+    [DataField("emagImmuneTag", customTypeSerializer: typeof(PrototypeIdSerializer<TagPrototype>))]
+    public string EmagImmuneTag = "EmagImmune";
+}
+
+/// <summary>
+/// Component for stunning mobs on click, when gloves are enabled.
+/// Knocks them down for a bit and deals shock damage.
+/// </summary>
+[RegisterComponent]
+public sealed class NinjaStunComponent : Component
```

Comment:
Same with this one, name it like "StunProviderComponent" or something and make it store the ChargeUsage, if given use power and make it stop working if power reaches 0, if not given make it work unpowered, also damage should be handled by the melee weapon component and *maybe* there is a component somewhere for StunTime (*maybe* stamina damage??)

Reply:
if it had a meleecomponent with boxing glove-like stats, wouldnt just punching someone with gloves work even if not enabled or as a ninja?

Reply:
Yeah thats a bug, tho it looks easily fixable tho, change
```cs
        if (EntityManager.TryGetComponent(entity, out HandsComponent? hands) &&
            hands.ActiveHandEntity is { } held)
        {
            if (EntityManager.TryGetComponent(held, out melee))
            {
                weaponUid = held;
                return true;
            }

            return false;
        }
```
in SharedWeaponsSystem.cs to
```cs
        if (TryComp(entity, out HandsComponent? hands) &&
            hands.ActiveHandEntity is { } held)
        {
            if (TryComp<ClothingComponent>(held, out var clothing) &&
                // new field, set it to gloves in the ninja gloves melee comp
                clothing.Slots.HasFlag(melee.RequiredSlot))
            {
                return false;
            }

            if (TryComp(held, out melee))
            {
                weaponUid = held;
                return true;
            }

            return false;
        }
```
(do test 4 me but it should work)

Reply:
looks like that would work for boxing gloves, but for ninja gloves it still:
- works while disabled
- doesnt need you to be a ninja
this would essentially require manually adding/removing MeleeWeaponComponent whenever state changes which is giga sus

Reply:
I mean, i think the gloves should always work as long as they are powered, but if you want to keep the action you can just make it add/remove the new components when toggled.
To make it ninja exclusive, if you implement these as i recommended, then they would be unusable to anyone that is not a ninja since ninja code is the only thing providing `component.ProviderUid`, thus if anyone else tried to use them they couldnt get power to reach the gloves in any way

Reply:
moved to stunprovider without melee weapon, just electrocting